### PR TITLE
HPCC-21233 Provide script for tighter HPCC/Nagios integration

### PR DESCRIPTION
**Pull request by eorochena**
•	Add the ability to monitor not default ports for dali, sasha, thor and esp services
•	Provides support for centralized nagios scenarios
Signed-off-by: Eduardo Orochena Eduardo.Orochena@lexisnexisrisk.com

### CODE REVIEW

**Comment by richardkchapman** [comment]
@wangkx Can you review this?

**Comment by xwang2713** [comment]
I will review it.

File: hpcc-nagios-tools/hpcc_centralized_nagios.py [hunk 1]
```diff
@@ -0,0 +1,707 @@
...
+        return ip_addresses
+
+    def NumberOfServers(self):
+        ParseXML = ET.parse(xml_local)
+        doc_root = ParseXML.getroot()
+        ip_addresses = []
+        for i in doc_root.findall('.//DafilesrvProcess/Instance'):
+            ip = i.attrib['netAddress']
+            ip_addresses.append(ip)
+        return ip_addresses
```
Comment (xwang2713) [inline hpcc-nagios-tools/hpcc_centralized_nagios.py]: Should return the size of the list?

File: hpcc-nagios-tools/hpcc_centralized_nagios.py [hunk 2]
```diff
@@ -0,0 +1,707 @@
...
+
+    def dali(self):
+        ParseXML = ET.parse(xml_local)
+        doc_root = ParseXML.getroot()
+        for i in doc_root.findall('.//DaliServerProcess/Instance'):
+            DaliIP = i.attrib['netAddress']
+        return DaliIP
```
Comment (xwang2713) [inline hpcc-nagios-tools/hpcc_centralized_nagios.py]: All following probably can be kept just one function with a parameter of (component process name). If dali, etc we only want to return the last one just use the last index to reference it.

File: hpcc-nagios-tools/hpcc_centralized_nagios.py [hunk 3]
```diff
@@ -0,0 +1,707 @@
...
+            Machine = node.attrib['name']
+            IP = node.attrib['netAddress']
+            ip_addr[Machine] = IP
+        RoxieSlaves = []
+        for key in ip_addr:
+            if key not in Roxies:
+                continue
+            elif key in Roxies:
+                RoxieSlaves.append(ip_addr[key])
+        return RoxieSlaves
```
Comment (xwang2713) [inline hpcc-nagios-tools/hpcc_centralized_nagios.py]: Roxie, Thor Master, Thor Slaves and MySQLRepo can be grouped to another type of function with process name as input parameter.

**Comment by xwang2713** [comment]
I see all components default ports are used for Nagios service definition.  What if the port is not default?
Also for multiple cluster such as Thor, Roxie and Esp each cluster may has a different port if deployed as containers on the host. "configgen" does give both ip list and port. Not sure how it outputs multiple ports. Maybe we can defer this complex scenarios later.

File: hpcc-nagios-tools/hpcc_centralized_nagios.py (hunk 2 above)
Comment (eorochena) [reply hpcc-nagios-tools/hpcc_centralized_nagios.py]: I implemented the suggested changes, I just need to run a couple of tests.

**Comment by eorochena** [comment]
> I see all components default ports are used for Nagios service definition. What if the port is not default?
> Also for multiple cluster such as Thor, Roxie and Esp each cluster may has a different port if deployed as containers on the host. "configgen" does give both ip list and port. Not sure how it outputs multiple ports. Maybe we can defer this complex scenarios later.

If the default ports are not being used, I can determine what those are from the environment.xml file as shown below

```        for i in doc_root.findall('.//ThorCluster'):
            try:
                Computer = i.attrib['computer']
                Port = i.attrib['masterport']
                if Computer in d:
                    machine.append(Port)
            except KeyError:
                continue
            masterport.setdefault(d[Computer], [])
            masterport[d[Computer]].append(machine)
            machine = []
```

**Comment by xwang2713** [comment]
Yes, I did see you handle Thor master port. But how about other service, for example dali: "check_command        check_dali!7070!5000". When checking dali the default port is used.

**Comment by eorochena** [comment]
> Yes, I did see you handle Thor master port. But how about other service, for example dali: "check_command check_dali!7070!5000". When checking dali the default port is used.

I can implement that change for the Dali and Sasha components in the current pull request 

```
def hpcc_master_components(self, component_name):
        ParseXML = ET.parse(xml_local)
        doc_root = ParseXML.getroot()
        ip_port = []
        for i in doc_root.findall('.//{}ServerProcess/Instance'.format(component_name)):
            master_component = i.attrib['netAddress']
            ip_port.append(master_component)
        for i in doc_root.findall('.//{}ServerProcess/Instance'.format(component_name))
            port = i.attrib['port']
            ip_port.append(port)
        return ip_port
```

**Comment by eorochena** [comment]
Implemented changes to handle non default ports for Dali, Sasha and Esp services.

**Comment by xwang2713** [comment]
Looks good

**Comment by rpastrana** [comment]
@richardkchapman please merge

**Pull request merged**